presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughProject.java

```diff
@@ -0,0 +1,101 @@
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.optimizations.SymbolMapper;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Optional;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - Project (non-identity)
+ *       - Source other than Filter / TableScan
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project
+ *    - TopN
+ *       - Source
+ * </pre>
+ */
+public final class PushTopNThroughProject
+        implements Rule<TopNNode>
+{
+    private static final Capture<ProjectNode> PROJECT_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN()
+                    .with(source().matching(
+                            project()
+                                    .matching(projectNode -> !projectNode.isIdentity())
+                                    .capturedAs(PROJECT_CHILD)
+                                    .with(source().matching(node -> !(node instanceof TableScanNode) && !(node instanceof FilterNode)))));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode parent, Captures captures, Context context)
+    {
+        ProjectNode projectNode = captures.get(PROJECT_CHILD);
+
+        Optional<SymbolMapper> symbolMapper = symbolMapper(parent.getOrderingScheme().getOrderBy(), projectNode.getAssignments());
+        if (!symbolMapper.isPresent()) {
+            return Result.empty();
+        }
+
+        TopNNode mappedTopN = symbolMapper.get().map(parent, projectNode.getSource(), context.getIdAllocator().getNextId());
+
+        return Result.ofPlanNode(
+                projectNode.replaceChildren(ImmutableList.of(
+                        mappedTopN)));
+    }
+
+    private Optional<SymbolMapper> symbolMapper(List<Symbol> symbols, Assignments assignments)
+    {
+        SymbolMapper.Builder mapper = SymbolMapper.builder();
+        for (Symbol symbol : symbols) {
+            Expression expression = assignments.get(symbol);
+            if (!(expression instanceof SymbolReference)) {
+                return Optional.empty();
```

Comment:
what if TopN is using different identity or renaming symbols? Then this rule could work also.
Consider plan like:
```
- Top(count=X, order_by=[a2,b2])
  - Project(a1->a2, b1->b2, c2=a1+b1)
```

Reply:
The rule works in this case.
Only OrderBy symbols are restricted to be plainly mapped (I mean not computed) by the projection (a2 and b2 in your example). The check you cited above is done for OrderBy symbols only.
Other symbols produced by the projection (like c2) meet no restrictions.

I will add a test for a 'computing' projection.